futures-util/src/future/shared.rs

```diff
@@ -153,17 +154,33 @@ impl<Fut> Shared<Fut> where Fut: Future {
 
     /// Safety: callers must first ensure that `self.inner.state`
     /// is `COMPLETE`
-    unsafe fn clone_output(&self) -> Arc<Fut::Output> {
-        if let FutureOrOutput::Output(item) = &*self.inner.future_or_output.get() {
-            item.clone()
-        } else {
-            unreachable!()
+    unsafe fn take_output(&mut self) -> Fut::Output {
+        if let Some(inner) = Arc::get_mut(&mut self.inner) {
```

Comment:
To make this effective `Arc<Inner>` would need to be put into an Option so that after the future returns `Poll::Ready`, it destroys gives up its strong reference to Inner. Otherwise I think we'll rarely hit this branch.

Reply:
Unsure what you mean about `Option` wrapping but the point about rarely being hit is true. If one makes a 100 clones of a `Shared` instance then this branch won't get hit in 99 of them.

In that case this may be a pessimisation as the branch could be more expensive than cloning the value one extra time (101 times instead of 100).

Reply:
`get_mut` only returns `Some(_)` if the `Arc` is the last one that points to the value. What I mean is that it will not hit it in 100 of 100 cases if the other 99 `Shared` instances holding references to `Inner` weren't dropped before the 100th is polled.